src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java

```diff
@@ -0,0 +1,102 @@
+import com.minecolonies.api.colony.requestsystem.token.IToken;
+import com.minecolonies.api.util.Log;
+import com.minecolonies.coremod.colony.Colony;
+import com.minecolonies.coremod.colony.buildings.workerbuildings.BuildingWareHouse;
+import com.minecolonies.coremod.colony.requestsystem.requesters.BuildingBasedRequester;
+import com.minecolonies.coremod.colony.requestsystem.resolvers.core.AbstractWarehouseRequestResolver;
+import com.minecolonies.coremod.tileentities.TileEntityWareHouse;
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * ----------------------- Not Documented Object ---------------------
+ */
+public class WarehouseConcreteRequestResolver extends AbstractWarehouseRequestResolver
+{
+    public WarehouseConcreteRequestResolver(
+      @NotNull final ILocation location,
+      @NotNull final IToken<?> token)
+    {
+        super(location, token);
+    }
+
+    @Override
+    public boolean canResolveRequest(@NotNull final IRequestManager manager, final IRequest<? extends IDeliverable> requestToCheck)
+    {
+        if (requestToCheck.getRequester() instanceof BuildingBasedRequester)
+        {
+            final BuildingBasedRequester requester = ((BuildingBasedRequester) requestToCheck.getRequester());
+            final Optional<IRequester> building = requester.getBuilding(manager, requestToCheck.getRequester().getId());
+            if (building.isPresent() && building.get() instanceof BuildingWareHouse)
+            {
+                return false;
+            }
+        }
+
+        if (!manager.getColony().getWorld().isRemote)
+        {
+            if (!isRequestChainValid(manager, requestToCheck))
+            {
+                return false;
+            }
+
+            final Colony colony = (Colony) manager.getColony();
+            final Set<TileEntityWareHouse> wareHouses = getWareHousesInColony(colony);
+            wareHouses.removeIf(Objects::isNull);
+
+            try
+            {
+                final IDeliverable deliverable = requestToCheck.getRequest();
+                if(deliverable instanceof IConcreteDeliverable)
+                {
+                    for(final ItemStack possible : ((IConcreteDeliverable) deliverable).getRequestedItems())
+                    {
+                        final ItemStack stack = possible.copy();
+                        stack.setCount(requestToCheck.getRequest().getMinimumCount());
+                        return wareHouses.stream().anyMatch(wareHouse -> wareHouse.hasMatchingItemStackInWarehouse(stack));
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Log.getLogger().error(e);
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    protected boolean isRequestFromSelf(final IRequest<?> requestToCheck)
```

Comment:
Couldn't this and the getPriority go into the abstract?

Reply:
The getPriority can. This one is resolving which of the two descendent types to create.